src/ruby/puppetserver-lib/puppet/server/compiler.rb

```diff
@@ -8,17 +8,57 @@ def initialize
         set_server_facts
       end
 
+      # Compiles a catalog according to the spec provided from the
+      # request.
+      # @param [Hash] request_data details about the catalog to be compiled
+      # @return [Hash] containing either just the catalog or catalog and logs,
+      #                if capturing logs was enabled
       def compile(request_data)
+        options = request_data['options'] || {}
+        # Default to capturing errors and warnings from compiles
+        options['capture_logs'] = true unless options['capture_logs']
+
         processed_hash = convert_java_args_to_ruby(request_data)
 
-        node = create_node(processed_hash)
+        if options['capture_logs']
+          catalog, logs = capture_logs do
+            compile_catalog(processed_hash)
+          end
 
-        catalog = Puppet::Parser::Compiler.compile(node, processed_hash['job_id'])
-        catalog.to_data_hash
+          { catalog: catalog, logs: logs }
+        else
+          catalog = compile_catalog(request_data)
+          { catalog: catalog }
+        end
       end
 
       private
 
+      def compile_catalog(request_data)
+        node = create_node(request_data)
+        catalog = Puppet::Parser::Compiler.compile(node, request_data['job_id'])
+        catalog.to_data_hash
+      end
+
+      def capture_logs(&block)
+        logs = []
+        result = nil
+        log_dest = Puppet::Test::LogCollector.new(logs)
+        Puppet::Util::Log.with_destination(log_dest) do
+          result = yield
+        end
+
+        log_entries = logs.map do |log|
+          log.to_data_hash
+        end.select do |log|
+          # Filter out debug messages, which may be verbose and
+          # contain sensitive data
+          log['level'] == 'warning' || log['level'] == 'error'
```

Comment:
What happens when compilation errors? Are there different kinds of errors? I was just thinking that there's bound to be some errors that raise rather than log, or is that incorrect? Is it ever valid to return nothing from the endpoint (besides a 500)? Would we want to return the logs but not a catalog? Not necessarily blockers but related questions I realized I don't know the answer to.

Reply:
I'll play around with the error case.

Reply:
I think most parsing and compilation errors become 500s, we seem to use `fail` a lot in the compiler code around https://github.com/puppetlabs/puppet/blob/master/lib/puppet/pops/issues.rb.

Which means that nothing is returned, not even logs. I'm not sure if there are other kinds of errors that I don't know how to trigger, but it still seems good capture `error` level messages, just in case. For sure warnings are sent out with this.

Reply:
@Magisus that `fail` is different from Ruby `fail` (yep, it is badly named). The `fail` in the evaluator/issue logic raises an exception.
Typically all exceptions are caught and investigated by the evaluator/compiler and they are possibly decorated with location information before re-raised (possibly as a different type of exception). Some errors/warnings (those that come from validation) can result in multiple logged issues. If there was just one, it is raised as the final cause of the failure, if there are many a "there were many issues" type of exception is raised at the end.

That leaves only catastrophic errors like OOM or out of stack that would just cause the runtime to crash.

Hope that helps.